File: 0002_Add_Two_Numbers/solution.go
package _002_Add_Two_Numbers

//* Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dumyNode := &ListNode{Val: 0}
	n1, n2, carry, current := 0, 0, 0, dumyNode
	for l1 != nil || l2 != nil || carry != 0 {
		if l1 == nil {
			n1 = 0
		} else {
			l1 = l1.Next
			n1 = l1.Val

		}
		if l2 == nil {
			n2 = 0
		} else {
			n2 = l2.Val
			l2 = l2.Next
		}
		current.Next = &ListNode{Val: (n1 + n2 + carry) % 10}
		current = current.Next
		carry = (carry + n1 + n2) / 10
	}
	return dumyNode.Next
}

File: 0002_Add_Two_Numbers/solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        temp = ListNode(0)
        l = temp
        carry = 0
        while l1 or l2:
            val1, l1 = (l1.val, l1.next) if l1 else (0, None)
            val2, l2 = (l2.val, l2.next) if l2 else (0, None)
            l.next = ListNode((carry + val1 + val2) % 10)
            l = l.next
            carry = (carry + val1 + val2) // 10
        if carry > 0:
            l.next = ListNode(carry)
        return l.next
